### ISSUE TITLE
Fix pre-commit hook to read staging area, instead of / in addition to working directory

### ISSUE THREAD

**Issue by matuskalas** [edamontology/edamontology#622]
The **pre-commit hook** at `.githooks/pre-commit` reads the `EDAM_dev.owl` file from the working directory, and not from the staging area.

This could lead to problems when an invalid `EDAM_dev.owl` file is staged, then fixed in the working directory but forgotten to stage, **then** the invalid file from the staging area can be commited.

The problem is illustrated in this history:        **(see suggestions for a solution further below this)**

_... replaced all `owl:Class` with `wol:Class`_ ...
```
Matus@Rip MINGW64 ~/Work/edamOntology/edamontology (experiment-with-pre-commit-hook)
$ git add EDAM_dev.owl

Matus@Rip MINGW64 ~/Work/edamOntology/edamontology (experiment-with-pre-commit-hook)
$ git commit -m "Replaced all owls🦉 with wols🦉"
INFO: Pre-commit hook running from /c/Users/Matus/Work/edamOntology/edamontology
git version 2.29.2.windows.2
0

Matus@Rip MINGW64 ~/Work/edamOntology/edamontology (experiment-with-pre-commit-hook)
$ git log
commit f3d8b283fb0373e086f5d52e25f24bb9677ceb1f (HEAD)
Author: Matúš Kalaš <matus.kalas@uib.no>
Date:   Sun Apr 4 18:05:22 2021 +0200

    Trivial change, updated date & time
```
_... commit failed (_pre-commit_ returned 0), the latest commit is still the previous one ..._
_... undo, i.e. replaced again all `wol:Class` back with `owl:Class` ..._
```
Matus@Rip MINGW64 ~/Work/edamOntology/edamontology (experiment-with-pre-commit-hook)
$ git diff
diff --git a/EDAM_dev.owl b/EDAM_dev.owl
index 3f41286..1640555 100644
--- a/EDAM_dev.owl
+++ b/EDAM_dev.owl
@@ -748,12 +748,12 @@
     <owl:ObjectProperty rdf:about="http://edamontology.org/has_topic">
         <owl:inverseOf rdf:resource="http://edamontology.org/is_topic_of"/>
         <rdfs:domain>
-            <wol:Class>
+            <owl:Class>
                 <owl:unionOf rdf:parseType="Collection">
                     <rdf:Description rdf:about="http://edamontology.org/data_0006"/>
                     <rdf:Description rdf:about="http://edamontology.org/operation_0004"/>
                 </owl:unionOf>
-            </wol:Class>
+            </owl:Class>
         </rdfs:domain>
         <rdfs:range rdf:resource="http://edamontology.org/topic_0003"/>
         <oboOther:is_anti_symmetric>false</oboOther:is_anti_symmetric>
@@ -939,12 +939,12 @@
     <owl:ObjectProperty rdf:about="http://edamontology.org/is_topic_of">
         <rdfs:domain rdf:resource="http://edamontology.org/topic_0003"/>
         <rdfs:range>
-            <wol:Class>
+            <owl:Class>
                 <owl:unionOf rdf:parseType="Collection">
                     <rdf:Description rdf:about="http://edamontology.org/data_0006"/>
                     <rdf:Description rdf:about="http://edamontology.org/operation_0004"/>
                 </owl:unionOf>
-            </wol:Class>
+            </owl:Class>
         </rdfs:range>
         <oboOther:is_anti_symmetric>false</oboOther:is_anti_symmetric>
         <oboOther:is_reflexive>false</oboOther:is_reflexive>
@@ -987,7 +987,7 @@

     <!-- http://edamontology.org/data_0005 -->

-    <wol:Class rdf:about="http://edamontology.org/data_0005">
+    <owl:Class rdf:about="http://edamontology.org/data_0005">
         <rdfs:subClassOf rdf:resource="http://www.w3.org/2002/07/owl#DeprecatedClass"/>
         <created_in>beta12orEarlier</created_in>
         <obsolete_since>beta12orEarlier</obsolete_since>
@@ -997,13 +997,13 @@
         <oboInOwl:inSubset rdf:resource="http://purl.obolibrary.org/obo/edam#obsolete"/>
         <rdfs:label>Resource type</rdfs:label>
         <owl:deprecated>true</owl:deprecated>
-    </wol:Class>
+    </owl:Class>



     <!-- http://edamontology.org/data_0006 -->

-    <wol:Class rdf:about="http://edamontology.org/data_0006">
+    <owl:Class rdf:about="http://edamontology.org/data_0006">
         <owl:disjointWith rdf:resource="http://edamontology.org/format_1915"/>
         <owl:disjointWith rdf:resource="http://edamontology.org/operation_0004"/>
         <owl:disjointWith rdf:resource="http://edamontology.org/topic_0003"/>
@@ -1025,7 +1025,7 @@
         <rdfs:seeAlso>http://wsio.org/data_002</rdfs:seeAlso>
         <rdfs:seeAlso>http://www.ifomis.org/bfo/1.1/snap#Continuant</rdfs:seeAlso>
         <rdfs:seeAlso>http://www.onto-med.de/ontologies/gfo.owl#Perpetuant</rdfs:seeAlso>
-    </wol:Class>

Matus@Rip MINGW64 ~/Work/edamOntology/edamontology (experiment-with-pre-commit-hook)
$ git commit -m "Fixed owls🦉🦉 again, at least I think so😉"
INFO: Pre-commit hook running from /c/Users/Matus/Work/edamOntology/edamontology
git version 2.29.2.windows.2
6955
[experiment-with-pre-commit-hook cd9999c] Fixed owls🦉🦉 again, at least I think so😉
 1 file changed, 6955 insertions(+), 6955 deletions(-)

Matus@Rip MINGW64 ~/Work/edamOntology/edamontology (experiment-with-pre-commit-hook)
$ git show
commit cd9999c058a02d5ce3629b76fb96d543794b9b06 (HEAD -> experiment-with-pre-commit-hook)
Author: Matúš Kalaš <matus.kalas@uib.no>
Date:   Sun Apr 4 18:14:25 2021 +0200

    Fixed owls🦉🦉 again, at least I think so😉

diff --git a/EDAM_dev.owl b/EDAM_dev.owl
index 1640555..3f41286 100644
--- a/EDAM_dev.owl
+++ b/EDAM_dev.owl
@@ -748,12 +748,12 @@
     <owl:ObjectProperty rdf:about="http://edamontology.org/has_topic">
         <owl:inverseOf rdf:resource="http://edamontology.org/is_topic_of"/>
         <rdfs:domain>
-            <owl:Class>
+            <wol:Class>
                 <owl:unionOf rdf:parseType="Collection">
                     <rdf:Description rdf:about="http://edamontology.org/data_0006"/>
                     <rdf:Description rdf:about="http://edamontology.org/operation_0004"/>
                 </owl:unionOf>
-            </owl:Class>
+            </wol:Class>
         </rdfs:domain>
         <rdfs:range rdf:resource="http://edamontology.org/topic_0003"/>
         <oboOther:is_anti_symmetric>false</oboOther:is_anti_symmetric>
@@ -939,12 +939,12 @@
     <owl:ObjectProperty rdf:about="http://edamontology.org/is_topic_of">
         <rdfs:domain rdf:resource="http://edamontology.org/topic_0003"/>
         <rdfs:range>
-            <owl:Class>
+            <wol:Class>
                 <owl:unionOf rdf:parseType="Collection">
                     <rdf:Description rdf:about="http://edamontology.org/data_0006"/>
                     <rdf:Description rdf:about="http://edamontology.org/operation_0004"/>
                 </owl:unionOf>
-            </owl:Class>
+            </wol:Class>
         </rdfs:range>
         <oboOther:is_anti_symmetric>false</oboOther:is_anti_symmetric>
         <oboOther:is_reflexive>false</oboOther:is_reflexive>
@@ -987,7 +987,7 @@

     <!-- http://edamontology.org/data_0005 -->

-    <owl:Class rdf:about="http://edamontology.org/data_0005">
+    <wol:Class rdf:about="http://edamontology.org/data_0005">
         <rdfs:subClassOf rdf:resource="http://www.w3.org/2002/07/owl#DeprecatedClass"/>
         <created_in>beta12orEarlier</created_in>
         <obsolete_since>beta12orEarlier</obsolete_since>
@@ -997,13 +997,13 @@
         <oboInOwl:inSubset rdf:resource="http://purl.obolibrary.org/obo/edam#obsolete"/>
         <rdfs:label>Resource type</rdfs:label>
         <owl:deprecated>true</owl:deprecated>
-    </owl:Class>
+    </wol:Class>



     <!-- http://edamontology.org/data_0006 -->

-    <owl:Class rdf:about="http://edamontology.org/data_0006">
+    <wol:Class rdf:about="http://edamontology.org/data_0006">
         <owl:disjointWith rdf:resource="http://edamontology.org/format_1915"/>
         <owl:disjointWith rdf:resource="http://edamontology.org/operation_0004"/>
         <owl:disjointWith rdf:resource="http://edamontology.org/topic_0003"/>
@@ -1025,7 +1025,7 @@
         <rdfs:seeAlso>http://wsio.org/data_002</rdfs:seeAlso>
:
```
_... oh no, the committed file is in fact invalid, with all the `wol:Class`es ..._

Ideally, it would be possible to validate the `EDAM_dev.owl` file in both the working directory, AND in the staging area (or at least in the staging area only). This could be accomplished by ONE of the following options (please decide when fixing this):
- Simple solution 1: `.githooks/pre-commit` checks only the staging area, and exits with 0 if no `owl:Class` found in the staged file.
- Simple solution 2: `.githooks/pre-commit` checks both the working directory, reports results, and then the staging area, to exit with 0 if no `owl:Class` found in the staged file.
- More complex solution: Create a separate validation script (let's call it _e.g._ `validate.sh`, can live in the root directory of the repo or inside a `/scripts` folder). This script will check the `EDAM_dev.owl` file and report back. Then update `.githooks/pre-commit` to run the `validate.sh` script on the **staged** `EDAM_dev.owl`, and return 0 if invalid.

**Comment by shreyaswikriti** [comment]
hello @matuskalas, I am an Outreachy Applicant looking for issues to solve. Can you suggest to me how to get started on this issue? I want to work on this.

**Comment by Uttam-Singhh** [comment]
Hello, @matuskalas I would also like to work on this as well, will try to find a simple solution first if it still not works then I will go for the Complex one.

**Comment by matuskalas** [comment]
Thank you for your interest @shreyaswikriti and @Uttam-Singhh! To solve this issue, you will need to:
 - know or learn some basic shell scripting (Bash)
 - find out how to read the staging area in Git

If any of this is too much of a challenge just like that, I can look for some good online guides and learning materials. But feel free to try yourselves first (or help each other) if you feel like ;-)

**Comment by shreyaswikriti** [comment]
thank you @matuskalas for the guidance. I am working on it.

**Comment by shreyaswikriti** [comment]
I have gone through the articles on the internet for the solution but nothing works for me. Kindly suggest some tutorials
![Screenshot 2021-04-05 234022](https://user-images.githubusercontent.com/57635473/113608520-53b3f580-9668-11eb-82ce-b72c6240bb05.png)

**Comment by CQW-Code** [comment]
I'm interested in guides and tutorials etc related to Bash scripts... don't know how far I'll get but looks interesting.

**Comment by matuskalas** [comment]
There is one simple solution to the same problem here: https://danilodellaquila.com/en/blog/use-git-stash-in-your-pre-commit-hook

And these are Bash shell tutorials I like and recommend:
- [Command line essentials for Bioinformaticians](https://bcc2020cle.bioinformatics.guide/) ([Slides here](https://rpubs.com/iracooke/bcc2020cle), [Video here](https://bcc2020.sched.com/event/c7Sz/command-line-essentials-for-bioinformaticians) or on [Vimeo](https://vimeo.com/441683367), Hands-on exercises linked from the first site)
- [Software Carpentry shell tutorial](http://swcarpentry.github.io/shell-novice/)

**Comment by Uttam-Singhh** [comment]
Thank you so much @matuskalas !

**Comment by matuskalas** [comment]
The basic problem is solved in #638 (https://github.com/edamontology/edamontology/commit/55170a7a130d4023143121c684b438431c00882d). For further work to make it more robust, there is issue #696.
